server/src/main/java/com/linecorp/centraldogma/server/internal/mirror/MirrorRunner.java

```diff
@@ -0,0 +1,149 @@
+
+import java.io.File;
+import java.util.Map;
+import java.util.Objects;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.LinkedBlockingQueue;
+import java.util.concurrent.ThreadPoolExecutor;
+import java.util.concurrent.TimeUnit;
+
+import com.google.common.base.MoreObjects;
+
+import com.linecorp.armeria.common.util.SafeCloseable;
+import com.linecorp.centraldogma.server.CentralDogmaConfig;
+import com.linecorp.centraldogma.server.command.CommandExecutor;
+import com.linecorp.centraldogma.server.internal.storage.project.ProjectApiManager;
+import com.linecorp.centraldogma.server.mirror.MirrorResult;
+import com.linecorp.centraldogma.server.mirror.MirroringServicePluginConfig;
+import com.linecorp.centraldogma.server.storage.repository.MetaRepository;
+
+import io.micrometer.core.instrument.MeterRegistry;
+import io.micrometer.core.instrument.binder.jvm.ExecutorServiceMetrics;
+import io.netty.util.concurrent.DefaultThreadFactory;
+
+public final class MirrorRunner implements SafeCloseable {
+
+    private final ProjectApiManager projectApiManager;
+    private final CommandExecutor commandExecutor;
+    private final File workDir;
+    private final MirroringServicePluginConfig mirrorConfig;
+    private final ExecutorService worker;
+
+    private final Map<MirrorKey, CompletableFuture<MirrorResult>> inflightRequests = new ConcurrentHashMap<>();
+
+    public MirrorRunner(ProjectApiManager projectApiManager, CommandExecutor commandExecutor,
+                        CentralDogmaConfig cfg, MeterRegistry meterRegistry) {
+        this.projectApiManager = projectApiManager;
+        this.commandExecutor = commandExecutor;
+        // TODO(ikhoon): Periodically clean up stale repositories.
+        workDir = new File(cfg.dataDir(), "_mirrors_manual");
+        MirroringServicePluginConfig mirrorConfig =
+                (MirroringServicePluginConfig) cfg.pluginConfigMap()
+                                                  .get(MirroringServicePluginConfig.class);
+        if (mirrorConfig == null) {
+            mirrorConfig = MirroringServicePluginConfig.INSTANCE;
+        }
+        this.mirrorConfig = mirrorConfig;
+
+        final ThreadPoolExecutor threadPoolExecutor = new ThreadPoolExecutor(
+                0, mirrorConfig.numMirroringThreads(),
+                // TODO(minwoox): Use LinkedTransferQueue when we upgrade to JDK 21.
+                60, TimeUnit.SECONDS, new LinkedBlockingQueue<>(),
+                new DefaultThreadFactory("mirror-api-worker", true));
+        threadPoolExecutor.allowCoreThreadTimeOut(true);
+        worker = ExecutorServiceMetrics.monitor(meterRegistry, threadPoolExecutor,
+                                                "mirrorApiWorker");
+    }
+
+    public CompletableFuture<MirrorResult> run(String projectName, String mirrorId) {
```

Comment:
Optional) For better UX, if a `mirrorRunId` and the `triggeredTime` is returned I think it may be more clear to users whether the mirror result is from an in-flight mirror or a new attempt.

Reply:
It is difficult to create a unique/incremental ID without an external system. `triggeredTime` can be easily added.

Reply:
I was imagining a simple UUID inside `MirrorKey`, but I guess just returning `triggeredTime` is fine too